#+OPTIONS: toc:nil num:nil author:nil todo:nil
#+EXCLUDE_TAGS: noexport

* TODO Next                                                        :noexport:

- Mapping should use fastq files (to make use of the quality) and maybe a
  mapping quality threshold should be applied.

- Verify how this is working and what would change with a gene on the - strand !!!

- See [[file:~/hub/allemand/gphn/gphn.org::*Final%20Workflow][Final Workflow]]

- CORNERCASE: Adding chromsize to be sure to be in the bounderies of
  chromosomes when checking splicing sites

- DEAL WITH ALTERNATIVE EXONS:

  - Alternative exons could be defined by only overlap (1 base overlap is
   sufficient I think) not containing like currently. Also The main exon should
   be the one most supported by reads (instead of longest).

  - For alternative exons and clustering, we could have a reciprocal blast to see
    if the best hit is the main exon or one of the alternative ones.

- Transcripts with less than 2 blast hits:

  - [X] Add a column "no_junctions" to junction composition tables

  - [ ] Add corresponding tests

    - Transcripts with a single exon hit

    - Transcripts with no exon hit at all
  
* TODO Biblio                                                      :noexport:

- See with EA:
  Implement criteria to further selected exons like in https://www.biorxiv.org/content/early/2018/02/05/260562.full.pdf
  - Looking for inserts in read mapping against transcriptome
  - Candidate exons from the mapping to genome should be 6nt away from known exon
  - Quantification by alignment to pseudo-transcriptome (downweight for multimappers), TMM norm.
    
* History
** DONE 0.4.0 Implement multiple fasta and alternative exons [4/4]
   CLOSED: [2019-08-08 Thu 16:09] SCHEDULED: <2019-08-16 Fri>

<2019-08-08 Thu>

- [X] Allow for multiple fasta files input and keep library information

- [X] Simplify interface and create scripts

- [X] Implement cryptic exons

- [X] Export all tables to csv

** DONE 0.3.0 Implement clustering and results export [2/2]
   CLOSED: [2019-07-24 Wed 09:48] SCHEDULED: <2019-07-24 Wed>

<2019-07-24 Wed>

- [X] Add TranscriptCluster object 
  
  > Running Blast identification, exon composition detection and data export as
  fasta and table.

- [X] Completing tests / docs

** DONE 0.2.0 Implement exon identification [3/3]
   CLOSED: [2019-07-19 Fri 16:50] SCHEDULED: <2019-07-19 Fri>

<2019-07-19 Fri>

- [X] Prepare test dataset
    
  > Using GPHN gene and previously described 40 exons

- [X] From multiple alignments of long reads against a genome, extract a list of putative exons as dataframe.

- [X] Export exons to bed and fasta files

